src/Components/Components/src/RenderTree/Renderer.cs

```diff
@@ -926,12 +932,15 @@ private void RenderInExistingBatch(RenderQueueEntry renderQueueEntry)
     {
         var componentState = renderQueueEntry.ComponentState;
         Log.RenderingComponent(_logger, componentState);
+        var startTime = (_renderingMetrics != null && _renderingMetrics.IsDurationEnabled()) ? Stopwatch.GetTimestamp() : 0;
+        _renderingMetrics?.RenderStart(componentState.Component.GetType().FullName);
         componentState.RenderIntoBatch(_batchBuilder, renderQueueEntry.RenderFragment, out var renderFragmentException);
         if (renderFragmentException != null)
         {
             // If this returns, the error was handled by an error boundary. Otherwise it throws.
             HandleExceptionViaErrorBoundary(renderFragmentException, componentState);
         }
+        _renderingMetrics?.RenderEnd(componentState.Component.GetType().FullName, renderFragmentException, startTime, Stopwatch.GetTimestamp());
```

Comment:
Several questions on this block:
* Do we understand what this is measuring?
  * This is measuring the `Render` method on the component, which just updates the RenderTree, which might not be very interesting.
  * It might make more sense to measure when a render batch starts and ends, as that more clearly represents the time it took the app to create a single "snapshot" update.
  * A more representative thing to measure is `SetComponentParametersAsync` which is where most of the application logic for an app lives.
* Do we know the perf cost of this? (Since it happens per render)
  * Does it make sense to cache the component `FullName` inside ComponentState

Reply:
>       * It might make more sense to measure when a render batch starts and ends
I agree that duration of whole batch could be separate metric.

>       * A more representative thing to measure is `SetComponentParametersAsync` which is where most of the application logic for an app lives.

OK. I will need some help to understand what are all the places in which this is called.
You mean `ComponentBase.SetParametersAsync()`, right ?

>     * Do we know the perf cost of this? (Since it happens per render)

It depends on if there is listener attached or not. When it's not it's negligible.

>       * Does it make sense to cache the component `FullName` inside ComponentState

Reflection already does some caching.